caddyhttp/proxy/reverseproxy.go

```diff
@@ -126,6 +128,35 @@ func NewSingleHostReverseProxy(target *url.URL, without string) *ReverseProxy {
 	return rp
 }
 
+func NewDynamicHostReverseProxy(templateUrl string, without string) *ReverseProxy {
+	director := func(req *http.Request) {
+		replacer := httpserver.NewReplacer(req, nil, "")
+
+		templateUrl = replacer.Replace(templateUrl)
+
+		target, _ := url.Parse(templateUrl)
+		targetQuery := target.RawQuery
+
+		req.URL.Scheme = target.Scheme
+		req.URL.Host = target.Host
+
+		req.URL.Path = singleJoiningSlash(target.Path, req.URL.Path)
+		if targetQuery == "" || req.URL.RawQuery == "" {
+			req.URL.RawQuery = targetQuery + req.URL.RawQuery
+		} else {
+			req.URL.RawQuery = targetQuery + "&" + req.URL.RawQuery
+		}
+
+		// We are then safe to remove the `without` prefix.
+		if without != "" {
+			req.URL.Path = strings.TrimPrefix(req.URL.Path, without)
+		}
+	}
+	rp := &ReverseProxy{Director: director, FlushInterval: 250 * time.Millisecond} // flushing good for streaming & server-sent events
+
+	return rp
+}
```

Comment:
This is the same thing as NewSingleHostReverseProxy, isn't it, but with a different director and without support for unix sockets?

Any chance we could reuse that code? i.e. call NewSingleHostReverseProxy somehow (maybe its signature has to change slightly?) and then just change the director?

Thanks for working through this. :) Not an easy change! But it's looking better already.

Once we get this looking good, it will need some tests...